### ISSUE TITLE
Mineflayer keeps crashing when bot trying to drop something out

### ISSUE THREAD

**Issue by Hazuki-san** [PrismarineJS/mineflayer#1872]
- [x] The [FAQ](https://github.com/PrismarineJS/mineflayer/blob/master/docs/FAQ.md) doesn't contain a resolution to my issue 

## Versions
 - mineflayer: 2.41.0
 - server: vanilla/spigot/paper 1.8.8
 - node: 15.2.1

## Detailed description of a problem
I keep getting this error,
```
TypeError: Cannot read property 'id' of undefined
    at confirmTransaction (C:\Mineflayer00\node_modules\mineflayer\lib\plugins\inventory.js:465:29)
    at Client.<anonymous> (C:\Mineflayer00\node_modules\mineflayer\lib\plugins\inventory.js:566:5)
    at Client.emit (node:events:329:20)
    at FullPacketParser.<anonymous> (C:\Mineflayer00\node_modules\minecraft-protocol\src\client.js:91:12)
    at FullPacketParser.emit (node:events:329:20)
    at addChunk (C:\Mineflayer00\node_modules\readable-stream\lib\_stream_readable.js:298:12)
    at readableAddChunk (C:\Mineflayer00\node_modules\readable-stream\lib\_stream_readable.js:280:11)
    at FullPacketParser.Readable.push (C:\Mineflayer00\node_modules\readable-stream\lib\_stream_readable.js:241:10)
    at FullPacketParser.Transform.push (C:\Mineflayer00\node_modules\readable-stream\lib\_stream_transform.js:139:32)
    at FullPacketParser._transform (C:\Mineflayer00\node_modules\protodef\src\serializer.js:89:10)
```

## What did you try yet?
This always happen if bot inventory is full. I tried to clean up the space (but not the armor), it is still crashing.

## Your current code
```js
/*
 * แจกฟรี ห้ามนำไปขาย
 * สำหรับเซิร์ฟเวอร์ mc-zero.net (GUI22)
 * Hazuki-san (Hoap) 2021
 * r0neko | Code Quality CHECK!
 */

// Packages
const util = require('util')
const async = require('async');
const math = require('mathjs');
const v = require('vec3');

// Config
const fs = require('fs')
let rawdata = fs.readFileSync('config.json');
let data = JSON.parse(rawdata);

// Setting stuff up
var host = data["ip"];
var port = data["port"];
var version = data["version"];
var altsfiles = data["alt_txt"];
var authme = data["authme_pw"];
var botowner = data["botowner"];
var interval = data["interval"];

// Mineflayer
const mineflayer = require('mineflayer')
const {
	pathfinder,
	Movements,
	goals: {
		GoalBlock,
		GoalNear
	}
} = require('mineflayer-pathfinder')

const readFile = (fileName) => util.promisify(fs.readFile)(fileName, 'utf8')
const wrap = module.exports.wrap = cb => new Promise(resolve => cb(resolve));

function makeBot(_u, ix) {
	return new Promise((resolve, reject) => {
		setTimeout(() => {
			const bot = mineflayer.createBot({
				username: _u,
				host: host,
				port: port,
				version: version
			})
			const mcData = require('minecraft-data')(bot.version)
			const defaultMove = new Movements(bot, mcData)
			console.log("Loaded: " + _u)

			function goGUI22() {
				setTimeout(bot.activateItem, 500);
				bot.on('windowOpen', async (window) => {
					window.requiresConfirmation = false // fix
					await bot.clickWindow(22, 0, 0)
				})
			}

			bot.loadPlugin(pathfinder)
			bot.once('spawn', () => {
				bot.chat('/login ' + authme)
				bot.pathfinder.setMovements(defaultMove)
				bot.pathfinder.setGoal(new GoalBlock(28, 68, 0))
				bot.on('goal_reached', (goal) => {
					// โดด เพราะเหมือนเซิร์ฟค้างหรือเปล่า? ไม่แน่ใจ
					bot.setControlState('jump', true)
					bot.setControlState('jump', false)
				})
			})

			bot.chatAddPattern(
				/(ตอนนี้กำลังเล่นเพลง)/,
				'nowplayingdetected',
				'we know that we logged in'
			)

			bot.chatAddPattern(
				/CHATGAME » ใครพิมพ์คำว่า (.*) เสร็จก่อนชนะ!/,
				'chatreactiondetected',
				'Yooo, isnt that some good shit right there'
			)

			bot.chatAddPattern(
				/ระบบได้ทำการลบเก็บขยะจำนวน (.*) ชิ้น/,
				'clearlagged',
				'ffs i hate myself',
			)

			const NowPlaying = () => {
				// รู้แล้วว่าอยู่ในล็อบบี้ งั้นไปกันเลย!
				setTimeout(goGUI22, 500);
			}

			bot.on('nowplayingdetected', NowPlaying)

			/*
			 * SNIPER
			*/
			var sniping = false;
			var snipe = function() {
				if (!sniping) return;
				bot.on('chatreactiondetected', matches => {
					bot.chat(`${matches}`) // ตรงๆแม่นๆ แน่นอนจริงๆ
				});
			};

			/*
			 * โค้ตตกปลา
			 */
			var fishing = false;
			var fish = function () {
				// ClearLag Fix
				if (!fishing) return;
				bot.on('clearlagged', matches => {
					bot.activateItem()
					bot.activateItem()
				});

				let running = true;
				(async () => {
					while (running) {
						await wrap(res => bot.equip(
							//bot.inventory.findInventoryItem(mcData.itemsByName.fishing_rod.id),
							bot.inventory.findInventoryItem(346),
							'hand',
							res
						));

						let bobber = await wrap(res => {
							/* @param {Entity} entity */
							let onSpawn = entity => {
								if (entity.objectType !== "Fishing Float") return
								bot.once('entitySpawn', onSpawn);
								res(entity);
							}
							bot.on('entitySpawn', onSpawn);
							bot.activateItem();
						});

						if (!running) {
							bot.activateItem();
							break;
						}

						// Soft <= 0.3
						// Medium <= 1.23
						// Rage <= 5
						// MAD < Infinity
						await wrap(res => {
							let onParticles = packet => {
								let pos = bobber.position
								//if (packet.particleId === 4 && packet.particles === 6 && pos.distanceTo(new v(packet.x, pos.y, packet.z)) <= 0.3) res();
								if (packet.particleId === 4 && packet.particles === 6) res();
								bot._client.once('world_particles', onParticles);
							}
							bot._client.once('world_particles', onParticles)
						});
						bot.activateItem();
					}
				})();
				return () => {
					running = false;
				}
			};

			/*
			 * โค้ตแสดงข้อความในเซิรฺ์ฟ
			 */
			bot.on('message', (cm) => {
				console.log(bot.username + ": " + cm.toString())
				if (cm.toString().includes(bot.username)) return
			})

			/*
			 * โค้ตแชท
			 */
			bot.on('chat', function (username, message) {
				var matches = botowner.filter(function(pattern) {
				  return new RegExp(pattern).test(username);
				})
				if (username == bot.username) return; //ถ้าตรงกับชื่อตัวเอง อย่าสนใจ
				if (!matches) return; // ถ้าไม่ใช่เจ้าของบอท อย่าสนใจ

				try {
					var result = math.eval(message);
					if (result != null && result != undefined)
						bot.chat(result + '');
				} catch (e) {}

				if (message.indexOf("me] !bot") >= 0) // ข้อความอันนี้มาจาก /w
					message = message.substring(4); // ตัดข้อความออก (คำว่า "me] ")
				else if (message.split(" ")[0].toLowerCase() == "!all") // ถ้าไม่ใช่ก็แสดงว่าแชทโลก
					message = ["!bot", ...message.split(" ").slice(1)].join(" "); // ลบคำว่า !all แล้วเปลี่ยนเป็น !bot แทน

				var args = message.split(' ');
				if (args[0].toLowerCase() == "!bot") {
					var command = args[1].toLowerCase();
					switch (command) {
						case 'sudo':
							if (args.slice(2).join(" ") == "")
								botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'จะให้พิมพ์ว่าอะไรหรอ 555');});
							bot.chat(args.slice(2).join(" "))
							break;
						case 'disconnect':
							bot.quit();
							break;
						case 'drop':
							try {
								bot.toss(parseInt(args[2]), null, (args[3]) ? parseInt(args[3]) : 1, function (err) {
									if (!err)
										botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'โยนของให้แล้วนะ');});
									else
										botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + err.message);});
								});
							} catch (e) {
								botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'ไม่รู้จักสิ่งนี้อ่ะ');});
							}
							break;
						case 'dump':
							var inventory = bot.inventory.slots;
							var items = [];
							for (var item in inventory) {
								if (!inventory[item]) continue;
								items.push({
									type: inventory[item].type,
									count: inventory[item].count
								});
							}
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + `กำลังโยนของทั้งหมด ${items.length} ออกจากตัว`);});
							async.forEachSeries(items, function (item, callback) {
								bot.toss(item.type, null, item.count, function (err) {
									if (err) console.log(err);
									callback();
								});
							});
							break;
						case 'equip':
							var slot = (['hand', 'head', 'torso', 'legs', 'feed'].indexOf(args[3]) < 0) ? 'hand' : args[3];
							bot.equip(bot.inventory.slots[parseInt(args[2])], slot, function (err) {
								if (!err) botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'ใส่ให้แล้วนะ');});
								else botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + err.message);});
							})
						case 'experience':
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'เลเวลตอนนี้: ' + bot.experience.level);});
							break;
						case 'fish':
							var startFishing = function () {
								fishing = true;
								fish();
								botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'เริ่มตกปลาแล้ว จะได้ปลาเยอะมั้ยนะ?');});
							};
							var stopFishing = function () {
								fishing = false;
								bot.activateItem();
								botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'เลิกตกปลาแล้ว');});
							};
							var toggleFishing = function () {
								if (!fishing) startFishing();
								else stopFishing();
							};

							if (args[2] == 'start') startFishing();
							else if (args[2] == 'stop') stopFishing();
							else toggleFishing();

							break;
						case 'snipe':
							var startSniping = function () {
								sniping = true;
								snipe();
								botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'พวกมันพิมพ์ไม่ทันแน่ 5555');});
							};
							var stopSniping = function () {
								sniping = false;
								botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'อะไรอ่ะ กำลังสนุกอยู่เลย ปิดซะล่ะ :<');});
							};
							var toggleSniping = function () {
								if (!sniping) startSniping();
								else stopSniping();
							};

							if (args[2] == 'start') startSniping();
							else if (args[2] == 'stop') stopSniping();
							else toggleSniping();

							break;
						case 'food':
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'หลอดอาหาร: ' + bot.food + '/20');});
							break;
						case 'health':
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'หลอดเลือด: ' + bot.health + '/20');});
							break;
						case 'help':
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'คำสั่งทั้งหมด: [ sudo, disconnect, drop, dump, equip, experience, fish, snipe, food, health ]')});
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'คำสั่งทั้งหมด (ต่อ): [ help, inventory, item_activate, item_deactivate, locate, look, ping ]')});
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'คำสั่งทั้งหมด (ต่อ): [ route_to, route_stop ]')});
							break;
						case 'inventory':
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + JSON.stringify(bot.inventory));});
							break;
						case 'item_activate':
							bot.activateItem();
							break;
						case 'item_deactivate':
							bot.deactivateItem();
							break;
						case 'locate':
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + `ตอนนี้เราอยู่ที่: ${JSON.stringify(bot.entity.position)}`);});
							break;
						case 'look':
							var dir = {
								yaw: parseFloat(args[2]),
								pitch: parseFloat(args[3])
							};
							if (dir.yaw == null || dir.pitch == null || isNaN(dir.yaw) || isNaN(dir.pitch))
								return botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'ไม่รู้จักเลขนี้นะ');});
							bot.look(dir.yaw, dir.pitch, true, function () {
								botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + `ตอนนี้กำลังมองที่ ${JSON.stringify(dir)}`);});
							});
							break;
						case 'ping':
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'เรายังอยู่ดีจ้า!');});
							break;
						case 'route_to':
							if (args[2] == 'me' || args[2] == 'here') {
								var target = bot.players[username].entity;
								bot.pathfinder.setMovements(defaultMove)
      							bot.pathfinder.setGoal(new GoalNear(target.position.x, target.position.y, target.position.z, 1))
								break;
							} else {
								var dest = {
									x: parseFloat(args[2]),
									y: parseFloat(args[3]),
									z: parseFloat(args[4])
								};
								if (dest.x == null || dest.y == null || dest.z == null || isNaN(dest.x) || isNaN(dest.y) || isNaN(dest.z))
									return botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'ที่นี่ที่ไหนหรอ?');});
								bot.pathfinder.setMovements(defaultMove)
								bot.pathfinder.setGoal(new GoalBlock(dest.x, dest.y, dest.z))
								break;
							}
							case 'route_stop':
								bot.pathfinder.setGoal(null)
								break;
							default:
								console.log(args)
								botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + 'ไม่รู้จักคำสั่งนี้');});
					}
				}
			});

			bot.on('kicked', (reason, loggedIn) => console.log(reason, loggedIn))
			bot.on('error', (err) => reject(err))
			setTimeout(() => reject(Error('Took too long to spawn.')), 5000) // 5 sec
		}, interval * ix)
	})
}


async function main() {
	const file = await readFile(altsfiles)
	const accounts = file.split(/\r?\n/)
	const botProms = accounts.map(makeBot)
	const bots = (await Promise.allSettled(botProms)).map(({
		value,
		reason
	}) => value || reason).filter(value => !(value instanceof Error))
	console.log(`All bots has successfully logged on!`)
}

main()
```

## Expected behavior
Should drop normally, not crashing.

## Additional context
None

**Comment by Hazuki-san** [comment]
Maybe this was caused by the server plugins? Like ServerListPlus?

**Comment by DrMoraschi** [comment]
Can you paste only the relevant part in the code?

**Comment by Hazuki-san** [comment]
> Can you paste only the relevant part in the code?

I doubt it is this part
```js
						case 'dump':
							var inventory = bot.inventory.slots;
							var items = [];
							for (var item in inventory) {
								if (!inventory[item]) continue;
								items.push({
									type: inventory[item].type,
									count: inventory[item].count
								});
							}
							botowner.forEach(function(ownerlist) { bot.chat('/w ' + ownerlist + ' ' + `กำลังโยนของทั้งหมด ${items.length} ออกจากตัว`);});
							async.forEachSeries(items, function (item, callback) {
								bot.toss(item.type, null, item.count, function (err) {
									if (err) console.log(err);
									callback();
								});
							});
							break;
```

**Comment by DrMoraschi** [comment]
Use bot.inventory.items() instead of slots

**Comment by Hazuki-san** [comment]
Solved for 'dump' command, this happens again when bot's inventory is full/equiping armor

**Comment by DrMoraschi** [comment]
If it's the same error, use the same solution. If not, can you send the code that's causing the error. Or alternatively you could join the Discord Server

**Comment by Hazuki-san** [comment]
I'll just join Discord server and discuss about it.

**Comment by williamd5** [comment]
> I'll just join Discord server and discuss about it.

Can you please not. What are the rest of us stuck with this issue supposed to do? Why is this closed if it's not a resolved issue?

**Comment by u9g** [comment]
> > I'll just join Discord server and discuss about it.
> 
> Can you please not. What are the rest of us stuck with this issue supposed to do? Why is this closed if it's not a resolved issue?

This is a duplicate of https://github.com/PrismarineJS/mineflayer/issues/1811

**Comment by williamd5** [comment]
> > > I'll just join Discord server and discuss about it.
> > 
> > 
> > Can you please not. What are the rest of us stuck with this issue supposed to do? Why is this closed if it's not a resolved issue?
> 
> This is a duplicate of #1811

Which is a duplicate of #1579 (an issue opened about half a year ago!). Should we expect to see this fixed at all?

**Comment by Hazuki-san** [comment]
> > > > I'll just join Discord server and discuss about it.
> > > 
> > > 
> > > Can you please not. What are the rest of us stuck with this issue supposed to do? Why is this closed if it's not a resolved issue?
> > 
> > 
> > This is a duplicate of #1811
> 
> Which is a duplicate of #1579 (an issue opened about half a year ago!). Should we expect to see this fixed at all?
> 
> EDIT: #1579 is also a closed, but seemingly unresolved issue. The person closing it says that there is "a clear solution" by just adding a check for `undefined`. Where do you add this check? What do you do if the variable is undefined (which is always)?

Well, to be honest... The issue doesn't resolved at all... This thing still happens and I'm unsure why it keeps happening

**Comment by u9g** [comment]
@williamd5 @Hazuki-san Would you all mind trying this branch to see if it fixes the problem you are having? https://github.com/PrismarineJS/mineflayer/pull/1976

**Comment by Hazuki-san** [comment]
> @williamd5 @Hazuki-san Would you all mind trying this branch to see if it fixes the problem you are having? #1976

I'll give them a try, thanks!
